common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.combat;
+
+import com.mojang.blaze3d.systems.RenderSystem;
+import com.mojang.blaze3d.vertex.BufferBuilder;
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.mojang.blaze3d.vertex.VertexConsumer;
+import com.wynntils.core.components.Models;
+import com.wynntils.core.config.Category;
+import com.wynntils.core.config.ConfigCategory;
+import com.wynntils.core.features.Feature;
+import com.wynntils.mc.event.PlayerRenderEvent;
+import com.wynntils.mc.event.TickEvent;
+import com.wynntils.models.character.type.ClassType;
+import com.wynntils.models.gear.type.GearInfo;
+import com.wynntils.models.gear.type.GearMajorId;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.colors.CustomColor;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.render.buffered.CustomRenderType;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.minecraft.client.gui.screens.inventory.InventoryScreen;
+import net.minecraft.client.renderer.MultiBufferSource;
+import net.minecraft.core.Position;
+import net.minecraft.world.entity.player.Player;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import org.joml.Matrix4f;
+
+@ConfigCategory(Category.COMBAT)
+public class RangeVisualizerFeature extends Feature {
+
+    private static final MultiBufferSource.BufferSource BUFFER_SOURCE =
+            MultiBufferSource.immediate(new BufferBuilder(256));
+    private static final int SEGMENTS =
+            128; // number of straight lines to draw when rendering circle, higher = smoother but more expensive
+    private static final float HEIGHT = 0.1f;
+    private final Map<Player, CustomColor> renderData = new HashMap<>();
+    private final List<Player> players = new ArrayList<>();
+
+    @SubscribeEvent
+    public void onPlayerRender(PlayerRenderEvent e) {
+        if (players.contains(e.getPlayer())) return;
+        players.add(e.getPlayer());
+
+        if (!renderData.containsKey(e.getPlayer())) return;
+
+        for (Map.Entry<Player, CustomColor> entry : renderData.entrySet()) {
```

Comment:
This will render all circles for all renderData, each time you come here to render one player.

Instead, do something like this:
```
Player player = e.getPlayer();
Pair<CustomColor, Integer> thisRenderData = renderData.get(player);
renderCircle(e.getPoseStack(), thisRenderData.b(), player.position(), thisRenderData.a());
```